File: src/use-cases/party/JoinParty.js
/**
 * JoinParty Use Case
 * Allows a user to join a party via invite code or party ID
 */

const PartyPlayer = require('../../domain/entities/PartyPlayer');
const logger = require('../../../logger');

class JoinParty {
    /**
     * @param {IPartyRepository} partyRepository - Party repository
     * @param {IUserRepository} userRepository - User repository
     */
    constructor(partyRepository, userRepository) {
        this.partyRepository = partyRepository;
        this.userRepository = userRepository;
    }

    /**
     * Execute the use case
     * @param {Object} request - Join request
     * @param {string} request.userId - User ID attempting to join
     * @param {string} [request.partyId] - Party ID (for public parties)
     * @param {string} [request.inviteCode] - Invite code (for private parties)
     * @returns {Promise<Object>} Join result with party and player info
     */
    async execute({ userId, partyId, inviteCode }) {
        try {
            // Validate input
            if (!userId || typeof userId !== 'string') {
                throw new Error('User ID is required');
            }

            if (!partyId && !inviteCode) {
                throw new Error('Either party ID or invite code is required');
            }

            // Verify user exists
            const user = await this.userRepository.findById(userId);
            if (!user) {
                throw new Error('User not found');
            }

            // Find party
            let party;
            if (inviteCode) {
                party = await this.partyRepository.findByInviteCode(inviteCode);
                if (!party) {
                    throw new Error('Invalid invite code');
                }
            } else {
                party = await this.partyRepository.findById(partyId);
                if (!party) {
                    throw new Error('Party not found');
                }

                // Check if party is public
                if (party.visibility !== 'public') {
                    throw new Error('Party is private. Use invite code to join.');
                }
            }

            // Check if party is finished
            if (party.status === 'finished') {
                throw new Error('Cannot join finished party');
            }

            // Check if party is full
            const players = await this.partyRepository.getPartyPlayers(party.id);
            if (players.length >= party.settings.playerCount) {
                throw new Error('Party is full');
            }

            // Check if user is already in party
            const existingPlayer = players.find(p => p.userId === userId);
            if (existingPlayer) {
                throw new Error('User is already in this party');
            }

            // Determine player index
            const playerIndex = players.length;

            // Create party player
            const partyPlayer = PartyPlayer.create(party.id, userId, playerIndex);

            // Add player to party
            const savedPlayer = await this.partyRepository.addPlayer(partyPlayer);

            logger.info('User joined party', {
                userId: userId,
                username: user.username,
                partyId: party.id,
                partyName: party.name,
                playerIndex: playerIndex
            });

            return {
                success: true,
                party: party.toPublicObject(),
                player: {
                    id: savedPlayer.id,
                    userId: savedPlayer.userId,
                    playerIndex: savedPlayer.playerIndex,
                    joinedAt: savedPlayer.joinedAt
                }
            };
        } catch (error) {
            logger.error('Join party error', {
                userId,
                partyId,
                inviteCode,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = JoinParty;

File: zapzap-rust/src/application/party/join_party.rs
use std::sync::Arc;

use super::seat::{seat_player, Seating};
use crate::domain::entities::{Party, PartyStatus, PartyVisibility};
use crate::domain::repositories::{PartyRepository, RepositoryError, UserRepository};

/// Join party input
pub struct JoinPartyInput {
    pub user_id: String,
    pub party_id: String,
    pub invite_code: Option<String>,
}

/// Join party output
pub struct JoinPartyOutput {
    pub party: Party,
    pub player_index: u8,
}

/// Join party use case
pub struct JoinParty<U: UserRepository, P: PartyRepository> {
    user_repo: Arc<U>,
    party_repo: Arc<P>,
}

impl<U: UserRepository, P: PartyRepository> JoinParty<U, P> {
    pub fn new(user_repo: Arc<U>, party_repo: Arc<P>) -> Self {
        Self {
            user_repo,
            party_repo,
        }
    }

    pub async fn execute(&self, input: JoinPartyInput) -> Result<JoinPartyOutput, JoinPartyError> {
        // Validate user exists
        let user = self
            .user_repo
            .find_by_id(&input.user_id)
            .await?
            .ok_or(JoinPartyError::UserNotFound)?;

        if user.is_bot() {
            return Err(JoinPartyError::Validation(
                "Bots cannot join parties directly".into(),
            ));
        }

        // Find party
        let party = self
            .party_repo
            .find_by_id(&input.party_id)
            .await?
            .ok_or(JoinPartyError::PartyNotFound)?;

        // A private party is joined with its invite code only (Node: JoinParty.js)
        if party.visibility == PartyVisibility::Private {
            match input.invite_code.as_deref() {
                Some(code) if code == party.invite_code => {}
                Some(_) => return Err(JoinPartyError::InvalidInviteCode),
                None => return Err(JoinPartyError::PrivateParty),
            }
        }

        // Check party status. Node checks a full party first (JoinParty.js), so a full
        // started party answers PARTY_FULL; Node lets anyone join a started party with
        // a free seat, which Rust refuses.
        if party.status != PartyStatus::Waiting {
            let players = self.party_repo.get_party_players(&party.id).await?;
            if party.is_full(players.len()) {
                return Err(JoinPartyError::PartyFull);
            }
            return Err(JoinPartyError::PartyNotWaiting);
        }

        // Check if already in party
        if self
            .party_repo
            .is_player_in_party(&input.party_id, &input.user_id)
            .await?
        {
            return Err(JoinPartyError::AlreadyInParty);
        }

        let player_index = match seat_player(&*self.party_repo, &party, &input.user_id).await? {
            Seating::Seated(seat) => seat,
            Seating::Full => return Err(JoinPartyError::PartyFull),
            Seating::AlreadyIn => return Err(JoinPartyError::AlreadyInParty),
        };

        Ok(JoinPartyOutput {
            party,
            player_index,
        })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum JoinPartyError {
    #[error("Validation error: {0}")]
    Validation(String),
    #[error("User not found")]
    UserNotFound,
    #[error("Party not found")]
    PartyNotFound,
    #[error("Party is private. Use invite code to join.")]
    PrivateParty,
    #[error("Invalid invite code")]
    InvalidInviteCode,
    #[error("Party is not in waiting state")]
    PartyNotWaiting,
    #[error("Already in party")]
    AlreadyInParty,
    #[error("Party is full")]
    PartyFull,
    #[error("Repository error: {0}")]
    Repository(#[from] RepositoryError),
}
